### PR TITLE
Refactor HyperOpt to use RayTune

### PR DESCRIPTION
**Pull request by jimthompson5802**
# Code Pull Requests

Changes:
* Retire "Serial" legacy Ludwig hyperopt executor 
* Consolidate on using RayTune hyperparameter specification
* Update relevant unit tests for the above changes

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -64,20 +64,18 @@ def hyperopt_results():
     hyperopt_configs = {
         "parameters": {
             "trainer.learning_rate": {
-                "type": "float",
-                "low": 0.0001,
-                "high": 0.01,
-                "space": "log",
-                "steps": 3,
```
Comment (w4nderlust) [inline tests/conftest.py:71]: it there no steps in raytune loguniform?

Comment (w4nderlust) [reply tests/conftest.py:71]: or maybe that's why qloguniform exists? (I have not looked deeply in raytune yet)

Comment (jimthompson5802) [reply tests/conftest.py:71]: @w4nderlust  converted to use `loguniform`.   After creating the PR, I realized there were a few other unit tests (`test_api.py` and `test_cli.py`) that invoked hyperopt functions.  Your observation was fixed to get these other unit tests to work.

File: ludwig/visualize.py [hunk 2]
```diff
@@ -3710,7 +3710,8 @@ def hyperopt_hiplot(hyperopt_stats_path, output_directory=None, **kwargs):
 
 def hyperopt_results_to_dataframe(hyperopt_results, hyperopt_parameters, metric):
     df = pd.DataFrame([{metric: res["metric_score"], **res["parameters"]} for res in hyperopt_results])
-    df = df.astype({hp_name: hp_params[TYPE] for hp_name, hp_params in hyperopt_parameters.items()})
+    # TODO: work out how to use SPACE key instead of TYPE key
+    # df = df.astype({hp_name: hp_params[SPACE] for hp_name, hp_params in hyperopt_parameters.items()})
```
Comment (w4nderlust) [inline ludwig/visualize.py:3714]: We can use a sort of inverse mapping of RAY_TUNE_FLOAT_SPACES and the like, Space -> type ?

Comment (jimthompson5802) [reply ludwig/visualize.py]: define function `visualize._convert_space_to_dtype()` to translate from SPACE to dtype.

File: ludwig/automl/base_config.py [hunk 3]
```diff
@@ -132,7 +132,7 @@ def _create_default_config(
     base_automl_config["hyperopt"]["executor"]["time_budget_s"] = time_limit_s
     if time_limit_s is not None:
         base_automl_config["hyperopt"]["sampler"]["scheduler"]["max_t"] = time_limit_s
-    base_automl_config["hyperopt"]["sampler"]["search_alg"]["random_state_seed"] = random_seed
+    base_automl_config["hyperopt"]["search_alg"]["random_state_seed"] = random_seed
```
Comment (w4nderlust) [inline ludwig/automl/base_config.py:135]: maybe use constants for HYPEROPT and SEARCH_ALG

File: ludwig/automl/utils.py [hunk 4]
```diff
@@ -126,7 +126,7 @@ def _add_transfer_config(base_config: Dict, ref_configs: Dict) -> Dict:
         point_to_evaluate = {}
         _add_option_to_evaluate(point_to_evaluate, min_dataset_config, hyperopt_params, COMBINER)
         _add_option_to_evaluate(point_to_evaluate, min_dataset_config, hyperopt_params, TRAINER)
-        base_config[HYPEROPT][SAMPLER]["search_alg"]["points_to_evaluate"] = [point_to_evaluate]
+        base_config[HYPEROPT]["search_alg"]["points_to_evaluate"] = [point_to_evaluate]
```
Comment (w4nderlust) [inline ludwig/automl/utils.py:129]: maybe use constants for HYPEROPT and SEARCH_ALG?

File: ludwig/hyperopt/execution.py [hunk 5]
```diff
@@ -696,13 +602,17 @@ def execute(
 
         mode = "min" if self.goal != MAXIMIZE else "max"
         metric = "metric_score"
-        if self.search_alg_dict is not None:
-            if TYPE not in self.search_alg_dict:
+        # if random seed not set, use Ludwig seed
+        self.search_algorithm.check_for_random_seed(random_seed)
+        if self.search_algorithm.search_alg_dict is not None:
+            if TYPE not in self.search_algorithm.search_alg_dict:
                 logger.warning("WARNING: Kindly set type param for search_alg " "to utilize Tune's Search Algorithms.")
```
Comment (w4nderlust) [inline ludwig/hyperopt/execution.py:609]: The warning message could be made more useful. We shoudl say that the search alg is not specified, that it should be set from the available ones (keys of the registry?) and which one it is set by default to.

File: ludwig/hyperopt/run.py [hunk 6]
```diff
@@ -175,7 +179,8 @@ def hyperopt(
     logger.info(pformat(hyperopt_config, indent=4))
     logger.info("\n")
 
-    sampler = hyperopt_config["sampler"]
+    sampler = hyperopt_config["sampler"]  # TODO: reserach how to refactor or eliminate
```
Comment (w4nderlust) [inline ludwig/hyperopt/run.py:182]: can we remove?

File: ludwig/hyperopt/run.py [hunk 7]
```diff
@@ -243,10 +248,10 @@ def hyperopt(
                 )
             )
 
-    hyperopt_sampler = get_build_hyperopt_sampler(sampler[TYPE])(goal, parameters, **sampler)
+    hyperopt_sampler = get_build_hyperopt_sampler(sampler[TYPE])(parameters)
```
Comment (w4nderlust) [inline ludwig/hyperopt/run.py:251]: can we remove it?

File: ludwig/hyperopt/execution.py [hunk 8]
```diff
@@ -184,121 +184,22 @@ def execute(
...
-            self.hyperopt_sampler.update_batch(zip(sampled_parameters, metric_scores))
-
-        ordered_trials = self.sort_hyperopt_results(trial_results)
-        return HyperoptResults(ordered_trials=ordered_trials)
-
-
 class RayTuneExecutor(HyperoptExecutor):
     def __init__(
         self,
         hyperopt_sampler,
```
Comment (w4nderlust) [inline ludwig/hyperopt/execution.py:190]: can we remove it?

File: ludwig/hyperopt/run.py [hunk 9]
```diff
@@ -358,7 +364,7 @@ def update_hyperopt_params_with_defaults(hyperopt_params):
     set_default_value(hyperopt_params, "metric", LOSS)
     set_default_value(hyperopt_params, "goal", MINIMIZE)
 
-    set_default_values(hyperopt_params[SAMPLER], {TYPE: "random"})
+    set_default_values(hyperopt_params[SAMPLER], {TYPE: "ray"})
```
Comment (w4nderlust) [inline ludwig/hyperopt/run.py:367]: can we remove it?

File: ludwig/automl/base_config.py (hunk 3 above)
Comment (jimthompson5802) [reply ludwig/automl/base_config.py:135]: done.

File: ludwig/automl/utils.py (hunk 4 above)
Comment (jimthompson5802) [reply ludwig/automl/utils.py:129]: done

File: ludwig/hyperopt/run.py (hunk 6 above)
Comment (jimthompson5802) [reply ludwig/hyperopt/run.py:182]: done

File: ludwig/hyperopt/execution.py (hunk 8 above)
Comment (jimthompson5802) [reply ludwig/hyperopt/execution.py:190]: @w4nderlust   Not sure I know the "it" in "can we remove it?"

File: ludwig/hyperopt/run.py (hunk 9 above)
Comment (jimthompson5802) [reply ludwig/hyperopt/run.py:367]: done

File: ludwig/hyperopt/execution.py (hunk 8 above)
Comment (w4nderlust) [reply ludwig/hyperopt/execution.py:190]: Weird, if you open up Files changed it points to the right place. Anyway, I was referring to `hyperopt_sampler` in `RayTuneExecutor`. As we are already passing the search algorithm maybe we can simplify and remove the sampler alltogether?

**Comment by amholler** [comment]
Hi, @jimthompson5802, Overall looks good!

In slack, I posted minor changes needed in this PR for sanity end-to-end AutoML; can you please add them?
In the meantime, I will use my copy of your PR with the sanity changes added to run some model accuracy runs.

File: ludwig/hyperopt/execution.py (hunk 8 above)
Comment (jimthompson5802) [reply ludwig/hyperopt/execution.py:190]: @w4nderlust   the answer to your question is the typical Software Engineer answer, "It depends." 😄 

The `hyperopt_sampler` is an instance of the [`RayTuneSampler` class](https://github.com/ludwig-ai/ludwig/blob/b8a180354ad965926f072df0f0221dc3a05442c5/ludwig/hyperopt/sampling.py#L67-L126).

Two methods that are needed in this class.  One is used internally in the object, the other one is invoked by the Executor.

In theory, we could migrate the methods to the `Hyperoptexecutor` superclass or its `RayTuneExecutor` subclass and then retire `RayTuneSampler` class.  This will probably take me 2 or 3 evenings to get everthing refactored and tested.

The one test I can't really do is the AutoML "getting the right answer" test.  This may mean @amholler  will need to redo the testing that is occuring right now.

Or we can leave it as is now and do the refactoring in a later release as part of code clean-up.  As best I can tell, this change is purely internal and does not affect any external facing APIs.  I don't think this will be a breaking change.

Let me know how you'd like me to proceed.

Comment (w4nderlust) [reply ludwig/hyperopt/execution.py:190]: Got it, makes sense. The way I think about it is that logically what the sampler was doing before was implementing a search algorithm, so now having the search algorithm itself, it feels like logically there shouldn't be a need for it, although I would need to look more closely at the entire code after refactoring to understand the implications of removing the sampler.

Anyway, because this would be a change that does not affect the interface, but an internal reorganization, maybe we can land this PR first, which unlocks Anne and others to adapt to the changes in the interface, and then consider doing a followup pr to remove cleanup the sampler. What do you think?

Comment (jimthompson5802) [reply ludwig/hyperopt/execution.py:190]: @w4nderlust   I think it is OK to leave it as is for now.  Maybe get a few other eyes on the sampler part and get feedback from others on the refactored hyperopt.  We can decide later on how much more code refactoring is needed.    So I'll mark this as resolved for now.  I'll focus on finishing up the doc updates in by this weekend.  this should put the PR in  a good place for the final merge of code and docs.

File: ludwig/hyperopt/sampling.py [hunk 10]
```diff
@@ -432,4 +130,166 @@ def get_build_hyperopt_sampler(strategy_type):
     return get_from_registry(strategy_type, sampler_registry)
 
 
-sampler_registry = {"grid": GridSampler, "random": RandomSampler, "pysot": PySOTSampler, "ray": RayTuneSampler}
+sampler_registry = {"ray": RayTuneSampler}
+
+
+# TODO: split to separate module?
```
Comment (justinxzhao) [inline ludwig/hyperopt/sampling.py:136]: +1 to moving this to a separate file, perhaps ludwig/hyperopt/search_algos.py

File: ludwig/utils/defaults.py [hunk 11]
```diff
@@ -161,16 +166,75 @@ def _upgrade_deprecated_fields(config: Dict[str, Any]):
             )
             feature[TYPE] = NUMBER
 
-    if HYPEROPT in config and "parameters" in config[HYPEROPT]:
-        hparams = config[HYPEROPT]["parameters"]
-        for k, v in list(hparams.items()):
-            substr = "training."
-            if k.startswith(substr):
+    if HYPEROPT in config:
```
Comment (justinxzhao) [inline ludwig/utils/defaults.py:169]: Thanks for adding this backward-compatibility code!

Could we add a test for `_upgrade_deprecated_fields` or `merge_with_defaults` that asserts on the full upgrade of a legacy config (that uses lots of these code paths)?

Such a test would be an easy source of reference for Ludwig users to see concretely what's changed.

**Review by justinxzhao — APPROVED** [review]
Thanks for these changes @jimthompson5802! Just a few nit-y comments from me.

**Comment by justinxzhao** [comment]
Merging for now to cut a 0.5.1 release. We can address minor comments in a follow-up PR.

File: ludwig/utils/defaults.py (hunk 11 above)
Comment (jimthompson5802) [reply ludwig/utils/defaults.py:169]: Address by PR https://github.com/ludwig-ai/ludwig/pull/2082

File: ludwig/hyperopt/sampling.py (hunk 10 above)
Comment (jimthompson5802) [reply ludwig/hyperopt/sampling.py:136]: Addressed by PR https://github.com/ludwig-ai/ludwig/pull/2082
